### PR TITLE
Add Signature for PrettyPrint

### PR DESCRIPTION
**Pull request by wildeng**
Added the prettyprint signature file.
Marked as WIP because I'm trying to write the related tests.

### CODE REVIEW

**Comment by wildeng** [comment]
Removing the WIP because I find difficult implementing the tests, although I will try more.

File: stdlib/prettyprint/0/prettyprint.rbs [hunk 1]
```diff
@@ -0,0 +1,364 @@
+# frozen_string_literal: true
```
Comment (soutaro) [inline stdlib/prettyprint/0/prettyprint.rbs:1]: RBS doesn't understand `frozen_string_literal` directive. ✂️

File: stdlib/prettyprint/0/prettyprint.rbs [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+  # This is a convenience method which is same as follows:
+  #
+  #     begin
+  #       q = PrettyPrint.new(output, maxwidth, newline, &genspace)
+  #       ...
+  #       q.flush
+  #       output
+  #     end
+  #
+  def self.format: (?untyped output, ?Integer maxwidth, ?String newline, ?untyped genspace) ?{ (Integer) -> Integer } -> untyped
```
Comment (soutaro) [inline stdlib/prettyprint/0/prettyprint.rbs:46]: I think we can make an interface with `<<` method, and use the interface as the type of `output` here?

File: stdlib/prettyprint/0/prettyprint.rbs [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+
+  # Breaks the buffer into lines that are shorter than #maxwidth
+  #
+    def break_outmost_groups: () -> untyped
+
+  # This adds `obj` as a text of `width` columns in width.
+  #
+  # If `width` is not specified, obj.length is used.
+  #
+  def text: (untyped obj, ?Integer width) -> untyped
```
Comment (soutaro) [inline stdlib/prettyprint/0/prettyprint.rbs:144]: I'm wondering if we need `untyped` here as the type of `obj`. Is there any reason to make it other than `String`?

File: stdlib/prettyprint/0/prettyprint.rbs [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+  #     #<PrettyPrint::Group:0x8354758 @depth=1, @breakables=[], @break=false>
+  #     #<PrettyPrint::Group:0x8354550 @depth=2, @breakables=[], @break=false>
+  #     #<PrettyPrint::Group:0x83541cc @depth=3, @breakables=[], @break=false>
+  #     #<PrettyPrint::Group:0x8347e54 @depth=4, @breakables=[], @break=false>
+  #
+  def current_group: () -> PrettyPrint::Group
+
+  # Breaks the buffer into lines that are shorter than #maxwidth
+  #
+    def break_outmost_groups: () -> untyped
```
Comment (soutaro) [inline stdlib/prettyprint/0/prettyprint.rbs:138]: Seems like these `-> untyped` can be `-> void`?

**Comment by soutaro** [comment]
@wildeng Thank you for opening this PR. 👏 

I post some comments about possible improvements. And hope this commit https://github.com/ruby/rbs/commit/021bb65308eb4fb7d02f4959d1034b5891b5eee5 helps you writing tests.

**Review by soutaro — CHANGES_REQUESTED** [review]
🚧

File: stdlib/prettyprint/0/prettyprint.rbs (hunk 1 above)
Comment (wildeng) [reply stdlib/prettyprint/0/prettyprint.rbs:1]: I will remove it

File: stdlib/prettyprint/0/prettyprint.rbs (hunk 4 above)
Comment (wildeng) [reply stdlib/prettyprint/0/prettyprint.rbs:138]: definitely. I will change this

File: stdlib/prettyprint/0/prettyprint.rbs (hunk 3 above)
Comment (wildeng) [reply stdlib/prettyprint/0/prettyprint.rbs:144]: This is a good point, although the documentation is a little bit obscure for me. It's implicit the use of String.

File: stdlib/prettyprint/0/prettyprint.rbs (hunk 2 above)
Comment (wildeng) [reply stdlib/prettyprint/0/prettyprint.rbs:46]: If I correctly understood something like: 

```
interface  _Output
  def <<: (?String obj, ?String sep, ?String newline, ?untyped genspace ) ?{ (Integer) -> Integer } -> void
end
```
An then the definition becomes

```
   def self.format: (?untyped output, ?Integer maxwidth, ?String newline, ?untyped genspace) ?{ (Integer) -> Integer } -> Output
```

**Comment by wildeng** [comment]
@soutaro thanks a lot for the comments and the test templates. I'm really happy to try to give back what Ruby gave to me in all these years.

File: stdlib/prettyprint/0/prettyprint.rbs (hunk 3 above)
Comment (soutaro) [reply stdlib/prettyprint/0/prettyprint.rbs:144]: Yeah, I also got confused a bit. 😓 
As far as I know, `obj` is (implicitly) assumed to be a `String`.

File: stdlib/prettyprint/0/prettyprint.rbs (hunk 2 above)
Comment (soutaro) [reply stdlib/prettyprint/0/prettyprint.rbs:46]: I mean like the following:

```rbs
class PrettyPrint
  interface _Output
    def <<: (String) -> void
  end

  def self.format: (?_Output output, ?Integer maxwidth, ?String newline, ?untyped genspace) ?{ (Integer) -> Integer } -> _Output # !?!?
end
```

Yes, it's my bad... I finally found that the type cannot be written in RBS... 😢 

Leave the type of the first argument `?untyped`. 🙏 (I'm sorry for the confusion.)

Comment (wildeng) [reply stdlib/prettyprint/0/prettyprint.rbs:46]: No worries at all :)

**Comment by wildeng** [comment]
I've checked the original code and modified the PR accordingly and did the requested changes. I also wrote the tests for it. 
I really hope it's better than when I opened it  😄

**Review by soutaro — APPROVED** [review]
Looks great! 💯 
Could you fixup the commits?

**Comment by wildeng** [comment]
> Looks great! 💯
> Could you fixup the commits?

Done :)

**Comment by soutaro** [comment]
Merged! Thank you! 🎉

**Pull request merged**